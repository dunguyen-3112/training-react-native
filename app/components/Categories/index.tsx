import { StyleSheet, FlatList, View } from 'react-native';
import React, { memo, useCallback, useState } from 'react';

import { COLORS } from '@constants';
import { ICategory } from '@types';
import { Button } from '@components/common';
import { CATEGORIES } from '@constants';

export interface ICategories {
  marginTop?: number;
  onSelect?: (ids: number[]) => void;
}

const Categories = ({ marginTop, onSelect }: ICategories) => {
  const [tags, setTags] = useState<number[]>([]);

  const handlePressTag = useCallback(
    (id: number) => {
      let newTags;
      if (tags.includes(id)) newTags = tags.filter((item) => item !== id);
      else newTags = [...tags, id];
      onSelect && onSelect(newTags);
      setTags(newTags);
    },
    [onSelect, tags]
  );

  const handleItemSeparatorComponent = useCallback(
    () => <View style={styles.item} />,
    []
  );

  const handleKeyExtractor = useCallback((item: ICategory) => item.id + '', []);

  const handleRenderItem = useCallback(
    ({ item }: { item: ICategory }) => {
      const { id, name } = item;

      const isActive = tags.includes(id);

      return (
        <Button
          label={name}
          type={isActive ? 'active' : 'default'}
          customStyle={{
            ...styles.tag,
            ...(isActive && { ...styles.active }),
          }}
          onPress={() => handlePressTag(id)}
        />
      );
    },
    [handlePressTag, tags]
  );

  return (
    <View style={[styles.container, { marginTop: marginTop }]}>
      <FlatList
        data={CATEGORIES}
        showsHorizontalScrollIndicator={false}
        horizontal
        ItemSeparatorComponent={handleItemSeparatorComponent}
        keyExtractor={handleKeyExtractor}
        renderItem={handleRenderItem}
      />
    </View>
  );
};

export default memo(Categories);

const styles = StyleSheet.create({
  container: {
    height: 26,
  },
  item: {
    marginLeft: 7,
  },
  tag: {
    borderRadius: 8,
    paddingVertical: 4,
    paddingHorizontal: 10,
  },
  active: {
    backgroundColor: 'rgba(28, 195, 121, 0.1)',
    borderWidth: 1,
    borderColor: COLORS.GREEN,
  },
});
