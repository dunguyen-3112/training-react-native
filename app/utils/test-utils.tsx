import { render, RenderOptions } from '@testing-library/react-native';
import React, { ReactElement } from 'react';
import { NavigationContainer } from '@react-navigation/native';

const AllTheProviders = ({ children }: { children: ReactElement }) => {
  return <NavigationContainer>{children}</NavigationContainer>;
};

const customRender = (ui: ReactElement, options?: RenderOptions) =>
  render(ui, { wrapper: AllTheProviders, ...options });

export * from '@testing-library/react-native';

export { customRender as render };
