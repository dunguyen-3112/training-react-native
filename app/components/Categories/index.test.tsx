import { CATEGORIES } from '@constants';
import { render, screen, fireEvent } from '@utils/test-utils';
import React from 'react';

import Categories from './index';

const mockOnPress = jest.fn();

const setup = () => {
  const component = render(<Categories />);
  expect(component.toJSON()).toMatchSnapshot();
};

describe('Categories', () => {
  it('should render List', () => {
    setup();
    CATEGORIES.forEach((category) => {
      expect(screen.getByText(category.name || '')).toBeTruthy();
    });
  });
});
