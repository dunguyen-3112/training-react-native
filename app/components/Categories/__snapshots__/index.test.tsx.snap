// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Categories should render List 1`] = `
<View
  style={
    Array [
      Object {
        "height": 26,
      },
      Object {
        "marginTop": undefined,
      },
    ]
  }
>
  <RCTScrollView
    ItemSeparatorComponent={[Function]}
    data={
      Array [
        Object {
          "id": 1,
          "name": "Breakfast",
        },
        Object {
          "id": 2,
          "name": "Dinner",
        },
        Object {
          "id": 3,
          "name": "Dry fruits",
        },
        Object {
          "id": 4,
          "name": "Lunch",
        },
        Object {
          "id": 5,
          "name": "Vegetables",
        },
        Object {
          "id": 6,
          "name": "Greenish",
        },
        Object {
          "id": 7,
          "name": "Non-Veg",
        },
        Object {
          "id": 8,
          "name": "Fast Food",
        },
      ]
    }
    getItem={[Function]}
    getItemCount={[Function]}
    horizontal={true}
    keyExtractor={[Function]}
    onContentSizeChange={[Function]}
    onLayout={[Function]}
    onMomentumScrollBegin={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    showsHorizontalScrollIndicator={false}
    stickyHeaderIndices={Array []}
    viewabilityConfigCallbackPairs={Array []}
  >
    <View>
      <View
        onLayout={[Function]}
        style={
          Array [
            Object {
              "flexDirection": "row",
            },
            null,
          ]
        }
      >
        <View
          accessible={true}
          collapsable={false}
          focusable={true}
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Array [
              Object {
                "alignItems": "center",
                "borderRadius": 8,
                "flexDirection": "row",
                "justifyContent": "center",
              },
              Object {
                "backgroundColor": "#ECECEC",
                "borderRadius": 8,
                "color": "#5C5C5C",
                "fontSize": 13,
                "fontWeight": "400",
                "paddingHorizontal": 10,
                "paddingVertical": 4,
              },
            ]
          }
        >
          <Text
            style={
              Object {
                "backgroundColor": "#ECECEC",
                "color": "#5C5C5C",
                "fontSize": 13,
                "fontWeight": "400",
              }
            }
          >
            Breakfast
          </Text>
        </View>
        <View
          style={
            Object {
              "marginLeft": 7,
            }
          }
        />
      </View>
      <View
        onLayout={[Function]}
        style={
          Array [
            Object {
              "flexDirection": "row",
            },
            null,
          ]
        }
      >
        <View
          accessible={true}
          collapsable={false}
          focusable={true}
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Array [
              Object {
                "alignItems": "center",
                "borderRadius": 8,
                "flexDirection": "row",
                "justifyContent": "center",
              },
              Object {
                "backgroundColor": "#ECECEC",
                "borderRadius": 8,
                "color": "#5C5C5C",
                "fontSize": 13,
                "fontWeight": "400",
                "paddingHorizontal": 10,
                "paddingVertical": 4,
              },
            ]
          }
        >
          <Text
            style={
              Object {
                "backgroundColor": "#ECECEC",
                "color": "#5C5C5C",
                "fontSize": 13,
                "fontWeight": "400",
              }
            }
          >
            Dinner
          </Text>
        </View>
        <View
          style={
            Object {
              "marginLeft": 7,
            }
          }
        />
      </View>
      <View
        onLayout={[Function]}
        style={
          Array [
            Object {
              "flexDirection": "row",
            },
            null,
          ]
        }
      >
        <View
          accessible={true}
          collapsable={false}
          focusable={true}
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Array [
              Object {
                "alignItems": "center",
                "borderRadius": 8,
                "flexDirection": "row",
                "justifyContent": "center",
              },
              Object {
                "backgroundColor": "#ECECEC",
                "borderRadius": 8,
                "color": "#5C5C5C",
                "fontSize": 13,
                "fontWeight": "400",
                "paddingHorizontal": 10,
                "paddingVertical": 4,
              },
            ]
          }
        >
          <Text
            style={
              Object {
                "backgroundColor": "#ECECEC",
                "color": "#5C5C5C",
                "fontSize": 13,
                "fontWeight": "400",
              }
            }
          >
            Dry fruits
          </Text>
        </View>
        <View
          style={
            Object {
              "marginLeft": 7,
            }
          }
        />
      </View>
      <View
        onLayout={[Function]}
        style={
          Array [
            Object {
              "flexDirection": "row",
            },
            null,
          ]
        }
      >
        <View
          accessible={true}
          collapsable={false}
          focusable={true}
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Array [
              Object {
                "alignItems": "center",
                "borderRadius": 8,
                "flexDirection": "row",
                "justifyContent": "center",
              },
              Object {
                "backgroundColor": "#ECECEC",
                "borderRadius": 8,
                "color": "#5C5C5C",
                "fontSize": 13,
                "fontWeight": "400",
                "paddingHorizontal": 10,
                "paddingVertical": 4,
              },
            ]
          }
        >
          <Text
            style={
              Object {
                "backgroundColor": "#ECECEC",
                "color": "#5C5C5C",
                "fontSize": 13,
                "fontWeight": "400",
              }
            }
          >
            Lunch
          </Text>
        </View>
        <View
          style={
            Object {
              "marginLeft": 7,
            }
          }
        />
      </View>
      <View
        onLayout={[Function]}
        style={
          Array [
            Object {
              "flexDirection": "row",
            },
            null,
          ]
        }
      >
        <View
          accessible={true}
          collapsable={false}
          focusable={true}
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Array [
              Object {
                "alignItems": "center",
                "borderRadius": 8,
                "flexDirection": "row",
                "justifyContent": "center",
              },
              Object {
                "backgroundColor": "#ECECEC",
                "borderRadius": 8,
                "color": "#5C5C5C",
                "fontSize": 13,
                "fontWeight": "400",
                "paddingHorizontal": 10,
                "paddingVertical": 4,
              },
            ]
          }
        >
          <Text
            style={
              Object {
                "backgroundColor": "#ECECEC",
                "color": "#5C5C5C",
                "fontSize": 13,
                "fontWeight": "400",
              }
            }
          >
            Vegetables
          </Text>
        </View>
        <View
          style={
            Object {
              "marginLeft": 7,
            }
          }
        />
      </View>
      <View
        onLayout={[Function]}
        style={
          Array [
            Object {
              "flexDirection": "row",
            },
            null,
          ]
        }
      >
        <View
          accessible={true}
          collapsable={false}
          focusable={true}
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Array [
              Object {
                "alignItems": "center",
                "borderRadius": 8,
                "flexDirection": "row",
                "justifyContent": "center",
              },
              Object {
                "backgroundColor": "#ECECEC",
                "borderRadius": 8,
                "color": "#5C5C5C",
                "fontSize": 13,
                "fontWeight": "400",
                "paddingHorizontal": 10,
                "paddingVertical": 4,
              },
            ]
          }
        >
          <Text
            style={
              Object {
                "backgroundColor": "#ECECEC",
                "color": "#5C5C5C",
                "fontSize": 13,
                "fontWeight": "400",
              }
            }
          >
            Greenish
          </Text>
        </View>
        <View
          style={
            Object {
              "marginLeft": 7,
            }
          }
        />
      </View>
      <View
        onLayout={[Function]}
        style={
          Array [
            Object {
              "flexDirection": "row",
            },
            null,
          ]
        }
      >
        <View
          accessible={true}
          collapsable={false}
          focusable={true}
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Array [
              Object {
                "alignItems": "center",
                "borderRadius": 8,
                "flexDirection": "row",
                "justifyContent": "center",
              },
              Object {
                "backgroundColor": "#ECECEC",
                "borderRadius": 8,
                "color": "#5C5C5C",
                "fontSize": 13,
                "fontWeight": "400",
                "paddingHorizontal": 10,
                "paddingVertical": 4,
              },
            ]
          }
        >
          <Text
            style={
              Object {
                "backgroundColor": "#ECECEC",
                "color": "#5C5C5C",
                "fontSize": 13,
                "fontWeight": "400",
              }
            }
          >
            Non-Veg
          </Text>
        </View>
        <View
          style={
            Object {
              "marginLeft": 7,
            }
          }
        />
      </View>
      <View
        onLayout={[Function]}
        style={
          Array [
            Object {
              "flexDirection": "row",
            },
            null,
          ]
        }
      >
        <View
          accessible={true}
          collapsable={false}
          focusable={true}
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Array [
              Object {
                "alignItems": "center",
                "borderRadius": 8,
                "flexDirection": "row",
                "justifyContent": "center",
              },
              Object {
                "backgroundColor": "#ECECEC",
                "borderRadius": 8,
                "color": "#5C5C5C",
                "fontSize": 13,
                "fontWeight": "400",
                "paddingHorizontal": 10,
                "paddingVertical": 4,
              },
            ]
          }
        >
          <Text
            style={
              Object {
                "backgroundColor": "#ECECEC",
                "color": "#5C5C5C",
                "fontSize": 13,
                "fontWeight": "400",
              }
            }
          >
            Fast Food
          </Text>
        </View>
      </View>
    </View>
  </RCTScrollView>
</View>
`;
